### PR TITLE
Feat/curveusd br strategy

### PR DESCRIPTION
**Pull request by majkic99**

### CODE REVIEW

File: test/strategies.js [hunk 1]
```diff
@@ -3814,6 +3814,149 @@ const createLiquityDsrSupplyStrategy = () => {
     return liquityDsrSupplyStrategy.encodeForDsProxyCall();
 };
 
+const createCurveUsdAdvancedRepayStrategy = () => {
+    const repayStrategy = new dfs.Strategy('CurveUsdAdvancedRepayStrategy');
+
+    repayStrategy.addSubSlot('&controllerAddress', 'address');
+    repayStrategy.addSubSlot('&ratioState', 'uint8');
+    repayStrategy.addSubSlot('&targetRatio', 'uint256');
+    repayStrategy.addSubSlot('&collToken', 'address');
```
Comment (yippee-ki-yay) [inline test/strategies.js:3823]: collAddress, crvUsdAddress for consistent naming

File: contracts/triggers/CurveUsdCollRatioTrigger.sol [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    /// @param ratio ratio that represents the triggerable point
+    /// @param state represents if we want the current state to be higher or lower than ratio param
+    struct SubParams {
+        address user;
+        address market;
+        uint256 ratio;
+        uint8 state;
+    }
+    
+    /// @dev checks current safety ratio of a CompoundV3 position and triggers if it's in a correct state
```
Comment (yippee-ki-yay) [inline contracts/triggers/CurveUsdCollRatioTrigger.sol:29]: CompoundV3?

**Review by yippee-ki-yay — APPROVED** [review]

File: contracts/actions/curveusd/helpers/CurveUsdHelper.sol [hunk 3]
```diff
@@ -4,16 +4,31 @@ pragma solidity =0.8.10;
...
 
-contract CurveUsdHelper is MainnetCurveUsdAddresses {
+contract CurveUsdHelper is MainnetCurveUsdAddresses, DSMath {
     using TokenUtils for address;
 
     error CurveUsdInvalidController();
 
     bytes4 constant CURVE_SWAPPER_ID = bytes4(keccak256("CurveUsdSwapper"));
 
+    function getCollateralRatio(address _user, address _controllerAddr) public view returns (uint256 collRatio, bool isInSoftLiquidation) {
```
Comment (rajko-z) [inline contracts/actions/curveusd/helpers/CurveUsdHelper.sol:19]: Would slightly refactor this function to make it more readable. Something like this


        ICrvUsdController crvController = ICrvUsdController(_controllerAddr);
        
        uint256 debt = crvController.debt(_user);
        if (debt == 0) return (0, false);
        
        (uint256 crvUsdCollAmount, uint256 collAmount) = getCollAmountsFromAMM(_controllerAddr, _user);
        isInSoftLiquidation = crvUsdCollAmount > 0;
        
        address amm = crvController.amm();
        address collToken = crvController.collateral_token();
        uint256 oraclePrice = ILLAMMA(amm).price_oracle();
        uint256 assetDec = IERC20(collToken).decimals();
        
        uint256 collAmountWAD = assetDec > 18 ? (collAmount / 10 ** (assetDec - 18)) : (collAmount * 10 ** (18 - assetDec));
        collRatio = wdiv(wmul(collAmountWAD, oraclePrice) + crvUsdCollAmount, debt);
    
But this is totally opinionated

File: scripts/deploy-curveUSD.js [hunk 4]
```diff
@@ -4,37 +4,57 @@
...
-    // console.log('CurveUsdView deployed to:', curveView.address);
-    // console.log('CurveUsdLevCreate deployed to:', curveUsedLevCreate.address);
-    // console.log('CurveUsdRepay deployed to:', curveUsdRepay.address);
-    console.log('CurveUsdSwapper deployed to:', curveUsdSwapper.address);
-    // console.log('CurveUsdSelfLiquidateWithColl deployed to:', curveUsdSelfLiquidateWithColl.address);
+    const curveusdrepay = await redeploy('CurveUsdRepay', addrs[network].REGISTRY_ADDR, true, true);
+    const curvetrigger = await redeploy('CurveUsdCollRatioTrigger', addrs[network].REGISTRY_ADDR, true, true);
+    const curvechecker = await redeploy('CurveUsdCollRatioCheck', addrs[network].REGISTRY_ADDR, true, true);
+    const curveswapper = await redeploy('CurveUsdSwapper', addrs[network].REGISTRY_ADDR, true, true);
+
```
Comment (rajko-z) [inline scripts/deploy-curveUSD.js:25]: minor: would rename to:
* curveUsdRepay
* curveUsdTrigger
* curveUsdChecker
* curveUsdSwapper

**Review by yippee-ki-yay — APPROVED** [review]

File: contracts/actions/curveusd/advanced/CurveUsdSwapper.sol [hunk 5]
```diff
@@ -225,30 +233,43 @@ contract CurveUsdSwapper is CurveUsdHelper, ExchangeHelper, GasFeeHelper, AdminA
...
+        // we need to take the fee for tx cost as well, as it"s in a strategy
+        
+        feeAmount += calcGasCost(_gasUsed, _token, 0);
+        
+        // gas fee can't go over 20% of the whole amount
+        if (feeAmount > (_destTokenAmount / 5)) {
+            feeAmount = _destTokenAmount / 5;
+        }
+    
+        if (_dfsFeeDivider != 0) {
```
Comment (kjaredic) [inline contracts/actions/curveusd/advanced/CurveUsdSwapper.sol:271]: if a user has a custom discount divider eq to 0 we wont take AUTOMATION_DFS_FEE here, is this expected or bug?

**Review by kjaredic — APPROVED** [review]

File: contracts/actions/curveusd/advanced/CurveUsdSwapper.sol (hunk 5 above)
Comment (majkic99) [reply contracts/actions/curveusd/advanced/CurveUsdSwapper.sol:271]: expected behaviour, so we dont need to add more params to curve usd swapper

**Review by yippee-ki-yay — APPROVED** [review]

**Pull request merged**